pkg/compiler/codegen.go

```diff
@@ -1159,14 +1172,45 @@ func (c *codegen) Visit(node ast.Node) ast.Visitor {
 }
 
 // processDefers emits code for `defer` statements.
+// TRY-related opcodes handle exception as follows:
+// 1. CATCH block is executed only if exception has occured.
+// 2. FINALLY block is always executed, but after catch block.
+// Go `defer` statements are a bit different:
+// 1. `defer` is always executed irregardless of whether an exception has occured.
+// 2. `recover` can or can not handle a possible exception.
+// Thus we use the following approach:
+// 1. Throwed exception is saved in a static field X, static fields Y and is set to true.
+// 2. CATCH and FINALLY blocks are the same, and both contain the same CALLs.
+// 3. In CATCH block we set Y to true and emit default return values if it is the last defer.
+// 4. Execute FINALLY block only if Y is false.
 func (c *codegen) processDefers() {
 	for i := len(c.scope.deferStack) - 1; i >= 0; i-- {
 		stmt := c.scope.deferStack[i]
 		after := c.newLabel()
 		emit.Jmp(c.prog.BinWriter, opcode.ENDTRYL, after)
+
+		c.setLabel(stmt.catchLabel)
+		emit.Opcode(c.prog.BinWriter, opcode.DROP) // drop exception, it will be pushed via `recover()`
```

Comment:
But where will `recover()` get it from if you're to drop it here?

Reply:
`recover()` return value from the static slot. This `DROP` is for value pushed by VM.
Actually we can try to set the slot here, instead of in `panic`.

Reply:
Think of non-Go contracts called from Go contract. We can get an exception from them and we should be able to recover.

Reply:
Yeah, fixed this.